crates/sui-indexer-alt-jsonrpc/src/api/coin.rs

```diff
@@ -0,0 +1,349 @@
+            coin_balance_bucket,
+        })) = page.cursor
+        {
+            query = query.filter(sql::<Bool>(&format!(
+                r#"ROW(candidates.coin_balance_bucket, candidates.cp_sequence_number, candidates.object_id)
+                 < ROW({coin_balance_bucket}, {cp_sequence_number}, '\x{object_id}'::bytea)"#,
+                object_id = hex::encode(object_id),
+            )));
+        }
+
+        // Finally we order by coin_balance_bucket and break ties by object_id.
+        query = query
+            .order_by(candidates!(coin_balance_bucket).desc())
+            .then_order_by(candidates!(cp_sequence_number).desc())
+            .then_order_by(candidates!(object_id).desc())
+            .limit(limit + 1);
+
+        #[derive(Queryable, Debug, Serialize, Deserialize)]
+        #[diesel(table_name = coin_balance_buckets)]
+        struct IdCpBalance {
+            object_id: Vec<u8>,
+            cp_sequence_number: i64,
+            coin_balance_bucket: Option<i16>,
+        }
+
+        let mut buckets: Vec<IdCpBalance> =
+            conn.results(query).await.context("Failed to query coins")?;
+
+        // Now gather pagination info.
+        let has_next_page = buckets.len() > limit as usize;
+        if has_next_page {
+            buckets.truncate(limit as usize);
+        }
+
+        let next_cursor = buckets
+            .last()
+            .map(|bucket| {
+                let object_id =
+                    ObjectID::from_bytes(&bucket.object_id).context("Failed to parse object id")?;
+                let balance_bucket = bucket
+                    .coin_balance_bucket
+                    .ok_or_else(|| anyhow::anyhow!("Coin balance bucket is null"))?;
+                Cursor(BalanceCursor {
+                    object_id,
+                    cp_sequence_number: bucket.cp_sequence_number,
+                    coin_balance_bucket: balance_bucket,
+                })
+                .encode()
+                .context("Failed to encode cursor")
+            })
+            .transpose()?;
+
+        let ids = buckets
+            .iter()
+            .map(|bucket| ObjectID::from_bytes(&bucket.object_id))
+            .collect::<Result<Vec<_>, _>>()
+            .context("Failed to parse object id")?;
+
+        Ok(JsonRpcPage {
+            data: ids,
+            next_cursor,
+            has_next_page,
+        })
+    }
+
+    async fn coin_response(&self, id: ObjectID) -> Result<Coin, RpcError<Error>> {
+        let Self(ctx, _) = self;
+        let stored_object = load_latest(ctx.loader(), id)
+            .await?
+            .ok_or_else(|| anyhow::anyhow!("Failed to load latest object {:?}", id))?;
+
+        let object: Object =
+            bcs::from_bytes(&stored_object.serialized_object.ok_or_else(|| {
+                anyhow::anyhow!("Failed to deserialize object {:?}", stored_object.object_id)
+            })?)
+            .context("Failed to deserialize object")?;
```

Comment:
#21161 includes a refactoring such that `load_latest` always returns an `Object`.